src/tests/end2end/CopyTests.cpp

```diff
@@ -0,0 +1,418 @@
+                
+                data.resize(texelCount);
+            }
+        };
+
+        SourceTexture CreateSourceTexture(unsigned int width, unsigned int height) {
+            uint32_t rowPitch = Align(kBytesPerTexel * width, kTextureRowPitchAlignment);
+
+            SourceTexture source(device, width, height);
+
+            FillTextureData(width, height, rowPitch / kBytesPerTexel, source.data.data());
+
+            nxt::Buffer buffer = utils::CreateFrozenBufferFromData(device, source.data.data(), static_cast<uint32_t>(sizeof(RGBA8) * source.data.size()), nxt::BufferUsageBit::TransferSrc);
+            nxt::CommandBuffer commands = device.CreateCommandBufferBuilder()
+                .TransitionTextureUsage(source.texture, nxt::TextureUsageBit::TransferDst)
+                .CopyBufferToTexture(buffer, 0, rowPitch, source.texture, 0, 0, 0, width, height, 1, 0)
+                .GetResult();
+
+            queue.Submit(1, &commands);
+
+            return source;
+        }
+
+        static void FillTextureData(uint32_t width, uint32_t height, uint32_t texelsPerRow, RGBA8* data) {
+            for (unsigned int y = 0; y < height; ++y) {
+                for (unsigned int x = 0; x < width; ++x) {
+                    unsigned int i = x + y * texelsPerRow;
+                    data[i] = RGBA8(x % 255, y % 255, 0, 255);
+                }
+            }
+        }
+
+        static void PackTextureData(const RGBA8* srcData, uint32_t width, uint32_t height, uint32_t srcTexelsPerRow, RGBA8* dstData, uint32_t dstTexelsPerRow) {
+            for (unsigned int y = 0; y < height; ++y) {
+                for (unsigned int x = 0; x < width; ++x) {
+                    unsigned int src = x + y * srcTexelsPerRow;
+                    unsigned int dst = x + y * dstTexelsPerRow;
+                    dstData[dst] = srcData[src];
+                }
+            }
+        }
+
+        static std::vector<RGBA8> PackedRegion(const RGBA8* data, uint32_t textureWidth, uint32_t x, uint32_t y, uint32_t width, uint32_t height) {
+            uint32_t rowPitch = Align(kBytesPerTexel * textureWidth, kTextureRowPitchAlignment);
+            std::vector<RGBA8> region(width * height);
+            PackTextureData(data + x + y * (rowPitch / kBytesPerTexel), width, height, rowPitch / kBytesPerTexel, region.data(), width);
+            return region;
+        }
+};
+
+class CopyTests_T2B : public CopyTests {
+    protected:
+
+        // Test that on initialization, the texture data is as expected
+        void TestFullTexture(uint32_t width, uint32_t height) {
+            SourceTexture source = CreateSourceTexture(width, height);
+            std::vector<RGBA8> expected = PackedRegion(source.data.data(), width, 0, 0, width, height);
+            EXPECT_TEXTURE_RGBA8_EQ(expected.data(), source.texture, 0, 0, width, height, 0) << "Failed initialization of " << width << "x" << height << " texture" << std::endl;
```

Comment:
It looks like a lot of the test relies on the internal implementation of EXPECT_TEXTURE_RGBA to do a texture to buffer copy with specific parameters. Can we remove this assumption and do the copy here directly?